### PR TITLE
[eas-cli] Support target specific entitlements

### PR DESCRIPTION
**Pull request by wkozyra95**
# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

Support target specific entitlements

# How

Resolve entitlements per the target
- When reading targets also check entitlements for each of them and store them in a target object
  - bare project: resolve from native code based on value of CODE_SGIN_ENTITLEMENTS i ntarget build settings
  - managed project: add entielemnts to target definition in app config `extra?.eas?.build?.experimental?.ios?.appExtensions`
- when syncing capabilities in credentials manager use per target values

# Test Plan

run build managed and bare with multiple targets and remote credentials

### CODE REVIEW

File: packages/eas-cli/src/project/ios/target.ts [hunk 1]
```diff
@@ -1,45 +1,120 @@
...
-  { buildConfiguration, buildScheme }: XcodeBuildContext
-): Promise<Target[]> {
+export async function resolveMangedProjectTargetsAsync({
+  exp,
+  projectDir,
+  xcodeBuildContext,
+  env,
+}: ResolveTargetOptions): Promise<Target[]> {
+  const { buildScheme, buildConfiguration } = xcodeBuildContext;
+  const applicationTarget = {
```
Comment (dsokal) [inline packages/eas-cli/src/project/ios/target.ts:46]: Add type.

File: packages/eas-cli/src/project/ios/target.ts [hunk 2]
```diff
@@ -1,45 +1,120 @@
...
+  const appExtensions: UserDefinedTarget[] =
+    exp.extra?.eas?.build?.experimental?.ios?.appExtensions ?? [];
+
+  const { error } = AppExtensionsConfigSchema.validate(appExtensions, {
+    allowUnknown: false,
+    abortEarly: false,
+  });
+  if (error) {
+    throw new Error(
+      `Failed to validate "extra.eas.build.experimental.ios.appExtensions" in you app config\n${error.message}`
```
Comment (dsokal) [inline packages/eas-cli/src/project/ios/target.ts:68]:
```suggestion
      `Failed to validate "extra.eas.build.experimental.ios.appExtensions" in you app config.\n${error.message}`
```

File: packages/eas-cli/src/project/ios/target.ts [hunk 3]
```diff
@@ -1,45 +1,120 @@
...
+    );
+  }
+
+  const extensionsTargets = appExtensions.map(extension => ({
```
Comment (dsokal) [inline packages/eas-cli/src/project/ios/target.ts:72]: Add type:
`const extensionsTargets: Target[]` ?

**Review by dsokal — COMMENTED** [review]
Can you update the PR description so it's more descriptive?

File: packages/eas-cli/src/credentials/ios/actions/SetUpBuildCredentials.ts [hunk 4]
```diff
@@ -50,7 +48,9 @@ export class SetUpBuildCredentials {
         Log.newLine();
       }
       const action = new SetUpTargetBuildCredentials({
-        ...this.options,
+        enterpriseProvisioning: this.options.enterpriseProvisioning,
+        distribution: this.options.distribution,
+        iosCapabilitiesOptions: { entitlements: target.entitlements },
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/credentials/ios/actions/SetUpBuildCredentials.ts:53]: all the code from `./crededentials/ios/appstore` is passing that as a wrapped object, so I didn't change that if eventually needs to be wrapped either way

File: packages/eas-cli/src/credentials/ios/utils/printCredentials.ts [hunk 5]
```diff
@@ -123,7 +123,7 @@ export function displayIosCredentials(
 export function displayProjectCredentials(
   app: App,
   appBuildCredentials: IosAppBuildCredentialsMap,
-  targets: Target[]
+  targets: Omit<Target, 'entitlements'>[]
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/credentials/ios/utils/printCredentials.ts:126]: that function is use when seting up credenitals from credenitals.json where we do not have entitlements, also entitlements are not needed for displaying credentilals

File: packages/eas-cli/src/project/ios/target.ts [hunk 6]
```diff
@@ -1,45 +1,120 @@
...
 import { XcodeBuildContext } from './scheme';
 
 interface UserDefinedTarget {
   targetName: string;
   bundleIdentifier: string;
   parentBundleIdentifier?: string;
+  entitlements: JSONObject;
+}
+
+interface ResolveTargetOptions {
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/project/ios/target.ts:23]: it's used in 3 places

File: packages/eas-cli/src/build/ios/build.ts [hunk 7]
```diff
@@ -36,13 +36,12 @@ export async function createIosContextAsync(
...
-      projectDir: ctx.projectDir,
-      exp: ctx.exp,
-    },
-    xcodeBuildContext
-  );
+  const targets = await resolveTargetsAsync({
+    projectDir: ctx.projectDir,
+    exp: ctx.exp,
+    xcodeBuildContext,
+    env: buildProfile.env ?? {},
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply packages/eas-cli/src/build/ios/build.ts:43]: I could but sooner or later `?? {}` will be necessary, are you saying to move it as late as possible?

[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/build/ios/build.ts:43]: Yes

File: packages/eas-cli/src/credentials/ios/actions/SetUpBuildCredentials.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (dsokal) [reply packages/eas-cli/src/credentials/ios/actions/SetUpBuildCredentials.ts:53]: I see, but is there ever anything else in the object?

**Review by dsokal — APPROVED** [review]

**Comment by wkozyra95** [comment]
/changelog-entry new-feature Support target specific entitlements.

**Review by dsokal — APPROVED** [review]

**Pull request merged**